src/librustc_driver/profile/trace.rs

```diff
@@ -0,0 +1,246 @@
+    assert!(cons.len() > 0 && cons[0] != "");
+    cons[0].to_string()
+}
+
+// First return value is text; second return value is a CSS class
+pub fn html_of_effect(eff: &Effect) -> (String, String) {
+    match *eff {
+        Effect::QueryBegin(ref qmsg, ref cc) => {
+            let cons = cons_of_query_msg(qmsg);
+            (cons.clone(),
+             format!("{} {}",
+                     cons,
+                     match *cc {
+                         CacheCase::Hit => "hit",
+                         CacheCase::Miss => "miss",
+                     }))
+        }
+    }
+}
+
+// First return value is text; second return value is a CSS class
+fn html_of_duration(_start: &Instant, dur: &Duration) -> (String, String) {
+    use rustc::util::common::duration_to_secs_str;
+    (duration_to_secs_str(dur.clone()),
+     "".to_string()
+    )
+}
+
+fn html_of_fraction(frac: f64) -> (String, String) {
+    let css = {
+        if       frac > 0.50  { format!("frac-50") }
+        else if  frac > 0.40  { format!("frac-40") }
+        else if  frac > 0.30  { format!("frac-30") }
+        else if  frac > 0.20  { format!("frac-20") }
+        else if  frac > 0.10  { format!("frac-10") }
+        else if  frac > 0.05  { format!("frac-05") }
+        else if  frac > 0.02  { format!("frac-02") }
+        else if  frac > 0.01  { format!("frac-01") }
+        else if  frac > 0.001 { format!("frac-001") }
+        else                  { format!("frac-0") }
+    };
+    let percent = frac * 100 as f64;
+    if percent > 0.1 as f64 { (format!("{:.1}%", percent), css) }
+    else { (format!("< 0.1%", ), css) }
+}
+
+fn total_duration(traces: &Vec<Rec>) -> Duration {
+    let mut sum : Duration = Duration::new(0,0);
+    for t in traces.iter() {
+        sum += t.duration;
+    }
+    return sum
+}
+
+fn duration_div(nom: Duration, den: Duration) -> f64 {
+    let nom_sec = nom.as_secs();
+    let den_sec = den.as_secs();
+    let nom_nanos = nom.subsec_nanos();
+    let den_nanos = den.subsec_nanos();
+    if nom_sec == den_sec {
+        if nom_sec == 0 {
+            nom_nanos as f64 / den_nanos as f64
+        } else {
+            panic!("FIXME(matthewhammer)")
+        }
+    } else {
+        panic!("FIXME(matthewhammer)")
+    }
+}
+
+fn write_traces_rec(file: &mut File, traces: &Vec<Rec>, total: Duration, depth: usize) {
+    for t in traces {
+        let (eff_text, eff_css_classes) = html_of_effect(&t.effect);
+        let (dur_text, dur_css_classes) = html_of_duration(&t.start, &t.duration);
+        let fraction = duration_div(t.duration, total);
+        let percent = fraction * 100 as f64;
+        let (frc_text, frc_css_classes) = html_of_fraction(fraction);
+        write!(file, "<div class=\"trace depth-{} extent-{}{} {} {} {}\">\n",
+               depth,
+               t.extent.len(),
+               /* Heuristic for 'important' CSS class: */
+               if t.extent.len() > 5 || percent >= 1.0 as f64 {
```

Comment:
Why do these `as f64` needed? `1.0` should be `f64` by default, or you could force the literal as f64 with `1.0_f64` (Similarly `100.0_f64` several lines above)